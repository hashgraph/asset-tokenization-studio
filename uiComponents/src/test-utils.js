import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { BasePlatformTheme } from "./Theme/theme";
var DefaultTheme = extendTheme(BasePlatformTheme);
export var render = function (ui, theme) {
    if (theme === void 0) { theme = DefaultTheme; }
    var wrapper = function (_a) {
        var children = _a.children;
        return (React.createElement(ChakraProvider, { theme: theme }, children));
    };
    return testingLibraryRender(ui, { wrapper: wrapper });
};
