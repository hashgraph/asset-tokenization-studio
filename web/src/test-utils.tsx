import React from "react";
import { ChakraProvider } from "@chakra-ui/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderOptions, RenderResult } from "@testing-library/react";
import { render, waitFor } from "@testing-library/react";
import type { MemoryHistory } from "history";
import { createMemoryHistory } from "history";
import { I18nextProvider } from "react-i18next";
import { Router } from "react-router-dom";
import i18n from "./i18n";
import theme from "./theme";
import userEvent from "@testing-library/user-event";

export const selectCalendar = async (
  component: RenderResult,
  id: string,
  day: string | number = new Date().getDate(),
) => {
  const calendar = component.getByTestId(id);
  await userEvent.click(calendar);

  await waitFor(() => {
    const daysToSelect = component.getAllByTestId(`day-${day}`);

    const dayToSelect = daysToSelect.find(
      (day) => !day.hasAttribute("disabled"),
    );

    if (dayToSelect) {
      userEvent.click(dayToSelect);
    }
  });

  await userEvent.click(document.body);
};

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: () => {},
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const memoryHistory = createMemoryHistory();

export const AllProviders = ({
  children,
  history = memoryHistory,
}: {
  children?: React.ReactNode;
  history?: MemoryHistory;
}) => {
  return (
    <I18nextProvider i18n={i18n}>
      <QueryClientProvider client={testQueryClient}>
        <ChakraProvider theme={theme}>
          <Router location={history.location} navigator={history}>
            {children}
          </Router>
        </ChakraProvider>
      </QueryClientProvider>
    </I18nextProvider>
  );
};

const customRender = (
  ui: React.ReactElement,
  {
    options,
    history,
  }: {
    options?: RenderOptions;
    history?: MemoryHistory;
  } = {},
): RenderResult =>
  render(ui, {
    wrapper: ({ children }) => (
      <AllProviders history={history}>{children}</AllProviders>
    ),
    ...options,
  });

export { customRender as render };
