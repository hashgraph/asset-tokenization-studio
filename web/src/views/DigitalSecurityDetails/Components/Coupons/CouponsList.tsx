import {
  CouponViewModel,
  GetAllCouponsRequest,
} from "@hashgraph/asset-tokenization-sdk";
import { useGetAllCoupons } from "../../../../hooks/queries/useCoupons";
import { useParams } from "react-router-dom";
import { createColumnHelper } from "@tanstack/table-core";
import { Table, Text } from "io-bricks-ui";
import { useTranslation } from "react-i18next";
import { COUPONS_FACTOR, DATE_TIME_FORMAT } from "../../../../utils/constants";
import { formatDate } from "../../../../utils/format";

export const CouponsList = () => {
  const { id } = useParams();

  const { t } = useTranslation("security", {
    keyPrefix: "details.coupons.list",
  });

  const {
    data: coupons,
    isLoading: isLoadingCoupons,
    isFetching: isFetchingCoupons,
  } = useGetAllCoupons(
    new GetAllCouponsRequest({
      securityId: id!,
    }),
  );

  const columnHelper = createColumnHelper<CouponViewModel>();

  const columns = [
    columnHelper.accessor("couponId", {
      header: t("columns.id"),
      enableSorting: true,
    }),
    columnHelper.accessor("recordDate", {
      header: t("columns.recordDate"),
      cell: (row) => formatDate(row.getValue(), DATE_TIME_FORMAT),
      enableSorting: false,
    }),
    columnHelper.accessor("executionDate", {
      header: t("columns.executionDate"),
      cell: (row) => formatDate(row.getValue(), DATE_TIME_FORMAT),
      enableSorting: false,
    }),
    columnHelper.accessor("rate", {
      header: t("columns.rate"),
      cell: (row) => `${parseInt(row.getValue()) / COUPONS_FACTOR}%`,
      enableSorting: false,
    }),
    columnHelper.accessor("snapshotId", {
      header: t("columns.snapshotId"),
      cell: (row) => row.getValue() ?? "-",
      enableSorting: false,
    }),
  ];

  return (
    <Table
      columns={columns}
      data={coupons ?? []}
      name="coupons-list"
      emptyComponent={<Text>{t("emptyTable")}</Text>}
      isLoading={isLoadingCoupons || isFetchingCoupons}
    />
  );
};
