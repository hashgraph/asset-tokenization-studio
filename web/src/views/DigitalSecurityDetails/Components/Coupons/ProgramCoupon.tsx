import { Button, Center, HStack, Stack, VStack } from "@chakra-ui/react";
import {
  CalendarInputController,
  InputNumberController,
} from "@hashgraph/securitytoken-uicomponents/Forms/Controllers";
import { Info } from "@phosphor-icons/react";
import { isAfterDate, min, required } from "../../../../utils/rules";
import { SubmitHandler, useForm } from "react-hook-form";
import { useTranslation } from "react-i18next";
import {
  GetBondDetailsRequest,
  SetCouponRequest,
} from "@hashgraph/securitytoken-sdk";
import { useParams } from "react-router-dom";
import { useCoupons } from "../../../../hooks/queries/useCoupons";
import { useGetBondDetails } from "../../../../hooks/queries/useGetSecurityDetails";
import { dateToUnixTimestamp } from "../../../../utils/format";
import { COUPONS_FACTOR } from "../../../../utils/constants";
import { PhosphorIcon, Text, Tooltip } from "@hashgraph/securitytoken-uicomponents";
import { isBeforeDate } from "../../../../utils/helpers";

interface ProgramCouponFormValues {
  rate: number;
  recordTimestamp: string;
  executionTimestamp: string;
}

export const ProgramCoupon = () => {
  const { mutate: createCoupon, isLoading } = useCoupons();
  const { control, formState, handleSubmit, watch, reset } =
    useForm<ProgramCouponFormValues>({
      mode: "all",
    });
  const { t: tForm } = useTranslation("security", {
    keyPrefix: "details.coupons.program.input",
  });
  const { t: tGlobal } = useTranslation("globals");
  const { id = "" } = useParams();
  const recordTimestamp = watch("recordTimestamp");

  const { data: bondDetails } = useGetBondDetails(
    new GetBondDetailsRequest({
      bondId: id,
    }),
  );

  const submit: SubmitHandler<ProgramCouponFormValues> = (params) => {
    const request = new SetCouponRequest({
      securityId: id ?? "",
      rate: (params.rate * COUPONS_FACTOR).toString(),
      recordTimestamp: dateToUnixTimestamp(params.recordTimestamp),
      executionTimestamp: dateToUnixTimestamp(params.executionTimestamp),
    });

    createCoupon(request, {
      onSuccess: () => {
        reset();
      },
    });
  };
  const canProgramCoupon = isBeforeDate(bondDetails?.maturityDate!)(new Date());

  return (
    <Center h="full">
      {canProgramCoupon ? (
        <VStack
          as="form"
          onSubmit={handleSubmit(submit)}
          w="500px"
          gap={6}
          py={6}
          data-testid="coupons-form"
        >
          <Stack w="full">
            <HStack justifySelf="flex-start">
              <Text textStyle="BodyTextRegularSM">
                {tForm("recordDate.label")}*
              </Text>
              <Tooltip label={tForm("recordDate.tooltip")} placement="right">
                <PhosphorIcon as={Info} />
              </Tooltip>
            </HStack>
            {bondDetails && (
              <CalendarInputController
                control={control}
                id="recordTimestamp"
                rules={{ required }}
                fromDate={new Date()}
                toDate={new Date(bondDetails.maturityDate)}
                placeholder={tForm("recordDate.placeholder")}
                withTimeInput
              />
            )}
          </Stack>
          <Stack w="full">
            <HStack justifySelf="flex-start">
              <Text textStyle="BodyTextRegularSM">
                {tForm("paymentDate.label")}*
              </Text>
              <Tooltip label={tForm("paymentDate.tooltip")} placement="right">
                <PhosphorIcon as={Info} />
              </Tooltip>
            </HStack>
            {bondDetails && (
              <CalendarInputController
                control={control}
                id="executionTimestamp"
                rules={{
                  required,
                  validate: isAfterDate(new Date(recordTimestamp)),
                }}
                fromDate={new Date()}
                toDate={new Date(bondDetails.maturityDate)}
                placeholder={tForm("paymentDate.placeholder")}
                withTimeInput
              />
            )}
          </Stack>
          <Stack w="full">
            <HStack justifySelf="flex-start">
              <Text textStyle="BodyTextRegularSM">{tForm("rate.label")}*</Text>
              <Tooltip label={tForm("rate.tooltip")} placement="right">
                <PhosphorIcon as={Info} />
              </Tooltip>
            </HStack>
            <InputNumberController
              autoFocus
              control={control}
              id="rate"
              rules={{ required, min: min(0) }}
              placeholder={tForm("rate.placeholder")}
              decimalScale={3}
              fixedDecimalScale={true}
              suffix="%"
              thousandSeparator=","
              decimalSeparator="."
            />
          </Stack>
          <Button
            data-testid="create-coupon-button"
            alignSelf="flex-end"
            data-tesdtid="send-button"
            isLoading={isLoading}
            isDisabled={!formState.isValid}
            type="submit"
          >
            {tGlobal("send")}
          </Button>
        </VStack>
      ) : (
        <Text textStyle="BodyRegularSM" data-testid="expired-warning">
          {tForm("expired")}
        </Text>
      )}
    </Center>
  );
};
