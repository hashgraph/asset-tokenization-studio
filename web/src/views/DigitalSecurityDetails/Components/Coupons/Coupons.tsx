import { Stack } from "@chakra-ui/react";
import { Tabs } from "@hashgraph/securitytoken-uicomponents/DataDisplay/Tabs";
import { useTranslation } from "react-i18next";
import { ProgramCoupon } from "./ProgramCoupon";
import { SeeCoupon } from "./SeeCoupon";

export const Coupons = () => {
  const { t: tTabs } = useTranslation("security", {
    keyPrefix: "details.coupons.tabs",
  });

  return (
    <Stack w="full" h="full" layerStyle="container">
      <Tabs
        tabs={[
          { content: <ProgramCoupon />, header: tTabs("program") },
          { content: <SeeCoupon />, header: tTabs("see") },
        ]}
        isFitted
      />
    </Stack>
  );
};
