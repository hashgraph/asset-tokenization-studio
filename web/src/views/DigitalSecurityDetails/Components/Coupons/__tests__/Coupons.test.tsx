import { Coupons } from "../Coupons";
import { render } from "../../../../../test-utils";

jest.mock("../../../../../hooks/queries/useGetSecurityDetails", () => ({
  ...jest.requireActual("../../../../../hooks/queries/useGetSecurityDetails"),
  useGetBondDetails: () => ({
    data: {
      currency: "0x858368",
      nominalValue: "50",
      startingDate: "2023-11-29T23:00:00.000Z",
      maturityDate: "2027-01-30T23:00:00.000Z",
    },
  }),
}));

describe(`${Coupons.name}`, () => {
  test("should render correctly", () => {
    const component = render(<Coupons />);

    expect(component.asFragment()).toMatchSnapshot();
  });
});
