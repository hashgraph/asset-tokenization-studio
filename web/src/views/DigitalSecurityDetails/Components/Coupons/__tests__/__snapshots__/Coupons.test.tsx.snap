// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coupons should render correctly 1`] = `
<DocumentFragment>
  <div
    class="chakra-stack css-lj1oo0"
  >
    <div
      class="chakra-tabs css-1bx5ylf"
    >
      <div
        aria-orientation="horizontal"
        class="chakra-tabs__tablist css-1lifx19"
        role="tablist"
      >
        <button
          aria-controls="tabs-:r0:--tabpanel-0"
          aria-selected="true"
          class="chakra-tabs__tab css-1uc8bdt"
          data-index="0"
          id="tabs-:r0:--tab-0"
          role="tab"
          tabindex="0"
          type="button"
        >
          List
        </button>
        <button
          aria-controls="tabs-:r0:--tabpanel-1"
          aria-selected="false"
          class="chakra-tabs__tab css-1uc8bdt"
          data-index="1"
          id="tabs-:r0:--tab-1"
          role="tab"
          tabindex="-1"
          type="button"
        >
          Program coupon
        </button>
        <button
          aria-controls="tabs-:r0:--tabpanel-2"
          aria-selected="false"
          class="chakra-tabs__tab css-1uc8bdt"
          data-index="2"
          id="tabs-:r0:--tab-2"
          role="tab"
          tabindex="-1"
          type="button"
        >
          See coupon
        </button>
        <button
          aria-controls="tabs-:r0:--tabpanel-3"
          aria-selected="false"
          class="chakra-tabs__tab css-1uc8bdt"
          data-index="3"
          id="tabs-:r0:--tab-3"
          role="tab"
          tabindex="-1"
          type="button"
        >
          Holders
        </button>
      </div>
      <div
        class="chakra-tabs__tab-indicator css-1axjpr8"
        style="position: absolute; transition-property: left, right, top, bottom, height, width; transition-duration: 0ms; transition-timing-function: cubic-bezier(0, 0, 0.2, 1); left: 0px; width: 0px;"
      />
      <div
        class="chakra-tabs__tab-panels css-1o421ra"
      >
        <div
          aria-labelledby="tabs-:r0:--tab-0"
          class="chakra-tabs__tab-panel css-1ip3kt8"
          id="tabs-:r0:--tabpanel-0"
          role="tabpanel"
          tabindex="0"
        >
          <div
            class="css-0"
          >
            <div
              class="css-68x580"
            >
              <table
                class="chakra-table css-y9vv58"
                data-testid="table-coupons-list"
              >
                <tbody
                  class="css-0"
                >
                  <tr
                    class="css-tlunve"
                  >
                    <th
                      class="css-u4onf3"
                    >
                      <div
                        class="css-1lekzkb"
                      >
                        <div
                          class="css-zvlevn"
                        >
                          ID
                        </div>
                        <button
                          aria-label="Sort"
                          class="chakra-button css-kdxsnu"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="chakra-icon css-ajfcdo"
                            focusable="false"
                            viewBox="0 0 16 16"
                          >
                            <path
                              d="M4.53817 5.19128C4.50027 5.09993 4.49031 4.99939 4.50956 4.90238C4.52881 4.80537 4.57639 4.71624 4.64629 4.64628L7.64629 1.64628C7.69273 1.59979 7.74787 1.56291 7.80857 1.53775C7.86927 1.51258 7.93433 1.49963 8.00004 1.49963C8.06575 1.49963 8.13081 1.51258 8.19151 1.53775C8.25221 1.56291 8.30736 1.59979 8.35379 1.64628L11.3538 4.64628C11.4238 4.7162 11.4715 4.80533 11.4908 4.90237C11.5101 4.99942 11.5002 5.10001 11.4624 5.19142C11.4245 5.28283 11.3603 5.36095 11.278 5.41589C11.1957 5.47082 11.099 5.5001 11 5.50003H5.00004C4.90115 5.50001 4.80449 5.47066 4.72228 5.41571C4.64006 5.36075 4.57599 5.28265 4.53817 5.19128Z"
                              fill="#505B6B"
                            />
                            <path
                              d="M5.00039 10.5H11.0004C11.0993 10.4999 11.1961 10.5292 11.2784 10.5841C11.3607 10.6391 11.4248 10.7172 11.4627 10.8086C11.5006 10.9 11.5105 11.0006 11.4912 11.0977C11.4718 11.1947 11.4241 11.2838 11.3541 11.3538L8.35414 14.3538C8.30771 14.4002 8.25256 14.4371 8.19186 14.4623C8.13116 14.4874 8.0661 14.5004 8.00039 14.5004C7.93469 14.5004 7.86962 14.4874 7.80892 14.4623C7.74822 14.4371 7.69308 14.4002 7.64664 14.3538L4.64664 11.3538C4.57664 11.2838 4.52895 11.1947 4.50963 11.0977C4.4903 11.0006 4.50021 10.9 4.53808 10.8086C4.57596 10.7172 4.64011 10.6391 4.7224 10.5841C4.8047 10.5292 4.90145 10.4999 5.00039 10.5Z"
                              fill="#505B6B"
                            />
                          </svg>
                        </button>
                      </div>
                    </th>
                    <th
                      class="css-u4onf3"
                    >
                      <div
                        class="css-1lekzkb"
                      >
                        <div
                          class="css-zvlevn"
                        >
                          Record Date
                        </div>
                      </div>
                    </th>
                    <th
                      class="css-u4onf3"
                    >
                      <div
                        class="css-1lekzkb"
                      >
                        <div
                          class="css-zvlevn"
                        >
                          Execution Date
                        </div>
                      </div>
                    </th>
                    <th
                      class="css-u4onf3"
                    >
                      <div
                        class="css-1lekzkb"
                      >
                        <div
                          class="css-zvlevn"
                        >
                          Coupon Rate
                        </div>
                      </div>
                    </th>
                    <th
                      class="css-u4onf3"
                    >
                      <div
                        class="css-1lekzkb"
                      >
                        <div
                          class="css-zvlevn"
                        >
                          Snapshot
                        </div>
                      </div>
                    </th>
                  </tr>
                  <tr
                    class="css-1wpwesf"
                    data-testid="row-0"
                  >
                    <td
                      class="css-1kbbn1r"
                    >
                      <div
                        class="css-zvlevn"
                      >
                        <div
                          class="chakra-skeleton css-yixgko"
                        />
                      </div>
                    </td>
                    <td
                      class="css-1kbbn1r"
                    >
                      <div
                        class="css-zvlevn"
                      >
                        <div
                          class="chakra-skeleton css-yixgko"
                        />
                      </div>
                    </td>
                    <td
                      class="css-1kbbn1r"
                    >
                      <div
                        class="css-zvlevn"
                      >
                        <div
                          class="chakra-skeleton css-yixgko"
                        />
                      </div>
                    </td>
                    <td
                      class="css-1kbbn1r"
                    >
                      <div
                        class="css-zvlevn"
                      >
                        <div
                          class="chakra-skeleton css-yixgko"
                        />
                      </div>
                    </td>
                    <td
                      class="css-1kbbn1r"
                    >
                      <div
                        class="css-zvlevn"
                      >
                        <div
                          class="chakra-skeleton css-yixgko"
                        />
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <div
          aria-labelledby="tabs-:r0:--tab-1"
          class="chakra-tabs__tab-panel css-1ip3kt8"
          hidden=""
          id="tabs-:r0:--tabpanel-1"
          role="tabpanel"
          tabindex="0"
        >
          <div
            class="css-l84sf8"
          >
            <form
              class="chakra-stack css-ip7fdm"
              data-testid="coupons-form"
            >
              <div
                class="chakra-stack css-1h0sumw"
              >
                <div
                  class="chakra-stack css-1nqqhpy"
                >
                  <p
                    class="css-0"
                  >
                    Record date*
                  </p>
                  <div
                    class="css-k008qs"
                  >
                    <svg
                      class="chakra-icon css-kpbtv0"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 256 256"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm16-40a8,8,0,0,1-8,8,16,16,0,0,1-16-16V128a8,8,0,0,1,0-16,16,16,0,0,1,16,16v40A8,8,0,0,1,144,176ZM112,84a12,12,0,1,1,12,12A12,12,0,0,1,112,84Z"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="chakra-form-control css-1kxonj9"
                  role="group"
                >
                  <label
                    class="css-1fey8w1"
                    for="popover-trigger-:r5:"
                  >
                    <div
                      class="chakra-input__group css-pwmwmr"
                      data-group="true"
                    >
                      <div
                        class="chakra-input__left-element css-irrpsr"
                      >
                        <svg
                          class="chakra-icon css-kpbtv0"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="0 0 256 256"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M208,32H184V24a8,8,0,0,0-16,0v8H88V24a8,8,0,0,0-16,0v8H48A16,16,0,0,0,32,48V208a16,16,0,0,0,16,16H208a16,16,0,0,0,16-16V48A16,16,0,0,0,208,32ZM72,48v8a8,8,0,0,0,16,0V48h80v8a8,8,0,0,0,16,0V48h24V80H48V48ZM208,208H48V96H208V208Z"
                          />
                        </svg>
                      </div>
                      <input
                        aria-controls="popover-content-:r5:"
                        aria-expanded="false"
                        aria-haspopup="dialog"
                        aria-label="Date"
                        autocomplete="off"
                        class="chakra-input css-p6x41e"
                        data-testid="recordTimestamp"
                        id="popover-trigger-:r5:"
                        name="recordTimestamp"
                        placeholder="Select record date"
                        value=""
                      />
                      <div
                        class="chakra-input__right-element css-1akgnmh"
                      >
                        <svg
                          aria-label="Select a date"
                          class="chakra-icon css-c6odlo"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="0 0 256 256"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
                          />
                        </svg>
                      </div>
                    </div>
                  </label>
                  <div
                    class="chakra-popover__popper css-1qq679y"
                    style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
                  >
                    <section
                      class="chakra-popover__content css-19djqmd"
                      id="popover-content-:r5:"
                      role="dialog"
                      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
                      tabindex="-1"
                    />
                  </div>
                </div>
              </div>
              <div
                class="chakra-stack css-1h0sumw"
              >
                <div
                  class="chakra-stack css-1nqqhpy"
                >
                  <p
                    class="css-0"
                  >
                    Payment date*
                  </p>
                  <div
                    class="css-k008qs"
                  >
                    <svg
                      class="chakra-icon css-kpbtv0"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 256 256"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm16-40a8,8,0,0,1-8,8,16,16,0,0,1-16-16V128a8,8,0,0,1,0-16,16,16,0,0,1,16,16v40A8,8,0,0,1,144,176ZM112,84a12,12,0,1,1,12,12A12,12,0,0,1,112,84Z"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="chakra-form-control css-1kxonj9"
                  role="group"
                >
                  <label
                    class="css-1fey8w1"
                    for="popover-trigger-:ra:"
                  >
                    <div
                      class="chakra-input__group css-pwmwmr"
                      data-group="true"
                    >
                      <div
                        class="chakra-input__left-element css-irrpsr"
                      >
                        <svg
                          class="chakra-icon css-kpbtv0"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="0 0 256 256"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M208,32H184V24a8,8,0,0,0-16,0v8H88V24a8,8,0,0,0-16,0v8H48A16,16,0,0,0,32,48V208a16,16,0,0,0,16,16H208a16,16,0,0,0,16-16V48A16,16,0,0,0,208,32ZM72,48v8a8,8,0,0,0,16,0V48h80v8a8,8,0,0,0,16,0V48h24V80H48V48ZM208,208H48V96H208V208Z"
                          />
                        </svg>
                      </div>
                      <input
                        aria-controls="popover-content-:ra:"
                        aria-expanded="false"
                        aria-haspopup="dialog"
                        aria-label="Date"
                        autocomplete="off"
                        class="chakra-input css-p6x41e"
                        data-testid="executionTimestamp"
                        id="popover-trigger-:ra:"
                        name="executionTimestamp"
                        placeholder="Select payment date"
                        value=""
                      />
                      <div
                        class="chakra-input__right-element css-1akgnmh"
                      >
                        <svg
                          aria-label="Select a date"
                          class="chakra-icon css-c6odlo"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="0 0 256 256"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M213.66,101.66l-80,80a8,8,0,0,1-11.32,0l-80-80A8,8,0,0,1,53.66,90.34L128,164.69l74.34-74.35a8,8,0,0,1,11.32,11.32Z"
                          />
                        </svg>
                      </div>
                    </div>
                  </label>
                  <div
                    class="chakra-popover__popper css-1qq679y"
                    style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
                  >
                    <section
                      class="chakra-popover__content css-19djqmd"
                      id="popover-content-:ra:"
                      role="dialog"
                      style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
                      tabindex="-1"
                    />
                  </div>
                </div>
              </div>
              <div
                class="chakra-stack css-1h0sumw"
              >
                <div
                  class="chakra-stack css-1nqqhpy"
                >
                  <p
                    class="css-0"
                  >
                    Coupon rate*
                  </p>
                  <div
                    class="css-k008qs"
                  >
                    <svg
                      class="chakra-icon css-kpbtv0"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="0 0 256 256"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm16-40a8,8,0,0,1-8,8,16,16,0,0,1-16-16V128a8,8,0,0,1,0-16,16,16,0,0,1,16,16v40A8,8,0,0,1,144,176ZM112,84a12,12,0,1,1,12,12A12,12,0,0,1,112,84Z"
                      />
                    </svg>
                  </div>
                </div>
                <div
                  class="chakra-form-control css-1kxonj9"
                  role="group"
                >
                  <label
                    class="css-1fey8w1"
                  >
                    <div
                      class="chakra-input__group css-pwmwmr"
                      data-group="true"
                    >
                      <input
                        class="chakra-input css-ifstuv"
                        data-testid="rate"
                        id="field-:rd:"
                        inputmode="numeric"
                        name="rate"
                        placeholder="0,123%"
                        type="text"
                        value=""
                      />
                    </div>
                  </label>
                </div>
              </div>
              <button
                class="chakra-button css-1i50xze"
                data-tesdtid="send-button"
                data-testid="create-coupon-button"
                disabled=""
                type="submit"
              >
                Send
              </button>
            </form>
          </div>
        </div>
        <div
          aria-labelledby="tabs-:r0:--tab-2"
          class="chakra-tabs__tab-panel css-1ip3kt8"
          hidden=""
          id="tabs-:r0:--tabpanel-2"
          role="tabpanel"
          tabindex="0"
        >
          <div
            class="css-3d6uxt"
          >
            <div
              class="chakra-stack css-owjkmg"
            >
              <form
                class="chakra-stack css-ip7fdm"
                data-testid="coupons-form"
              >
                <div
                  class="chakra-stack css-1h0sumw"
                >
                  <div
                    class="chakra-stack css-1nqqhpy"
                  >
                    <p
                      class="css-0"
                    >
                      Coupon ID*
                    </p>
                    <div
                      class="css-k008qs"
                    >
                      <svg
                        class="chakra-icon css-kpbtv0"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 256 256"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm16-40a8,8,0,0,1-8,8,16,16,0,0,1-16-16V128a8,8,0,0,1,0-16,16,16,0,0,1,16,16v40A8,8,0,0,1,144,176ZM112,84a12,12,0,1,1,12,12A12,12,0,0,1,112,84Z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    class="chakra-form-control css-1kxonj9"
                    role="group"
                  >
                    <label
                      class="css-1fey8w1"
                    >
                      <div
                        class="chakra-input__group css-pwmwmr"
                        data-group="true"
                      >
                        <input
                          class="chakra-input css-ifstuv"
                          data-testid="couponId"
                          id="field-:rg:"
                          inputmode="numeric"
                          name="couponId"
                          placeholder="Add ID"
                          type="text"
                          value=""
                        />
                      </div>
                    </label>
                  </div>
                </div>
                <div
                  class="chakra-stack css-1h0sumw"
                >
                  <div
                    class="chakra-stack css-1nqqhpy"
                  >
                    <p
                      class="css-0"
                    >
                      Account ID*
                    </p>
                    <div
                      class="css-k008qs"
                    >
                      <svg
                        class="chakra-icon css-kpbtv0"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 256 256"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm16-40a8,8,0,0,1-8,8,16,16,0,0,1-16-16V128a8,8,0,0,1,0-16,16,16,0,0,1,16,16v40A8,8,0,0,1,144,176ZM112,84a12,12,0,1,1,12,12A12,12,0,0,1,112,84Z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    class="chakra-form-control css-1kxonj9"
                    role="group"
                  >
                    <label
                      class="css-1fey8w1"
                      for="targetId"
                    >
                      <div
                        class="chakra-input__group css-pwmwmr"
                        data-group="true"
                      >
                        <input
                          class="chakra-input css-ifstuv"
                          data-testid="targetId"
                          id="targetId"
                          name="targetId"
                          placeholder="Add ID"
                          value=""
                        />
                      </div>
                    </label>
                  </div>
                </div>
                <button
                  class="chakra-button css-1i50xze"
                  data-tesdtid="send-button"
                  disabled=""
                  type="submit"
                >
                   Check
                </button>
              </form>
            </div>
          </div>
        </div>
        <div
          aria-labelledby="tabs-:r0:--tab-3"
          class="chakra-tabs__tab-panel css-1ip3kt8"
          hidden=""
          id="tabs-:r0:--tabpanel-3"
          role="tabpanel"
          tabindex="0"
        >
          <div
            class="css-1811skr"
          >
            <div
              class="chakra-stack css-1dpdux6"
            >
              <div
                class="chakra-form-control css-1kxonj9"
                role="group"
              >
                <label
                  class="css-1fey8w1"
                  for="couponId"
                >
                  <span
                    class="css-1o8l9xj"
                    id="label"
                  >
                    Coupon ID*
                  </span>
                  <div
                    class="chakra-input__group css-pwmwmr"
                    data-group="true"
                  >
                    <input
                      class="chakra-input css-ifstuv"
                      data-testid="couponId"
                      id="couponId"
                      name="couponId"
                      placeholder="1"
                      value=""
                    />
                  </div>
                </label>
              </div>
              <button
                class="chakra-button css-nqintf"
                disabled=""
                type="button"
              >
                Search
              </button>
            </div>
            <div
              class="css-0"
            >
              <div
                class="css-68x580"
              >
                <table
                  class="chakra-table css-y9vv58"
                  data-testid="table-coupons-holders"
                >
                  <tbody
                    class="css-0"
                  >
                    <tr
                      class="css-tlunve"
                    >
                      <th
                        class="css-u4onf3"
                      >
                        <div
                          class="css-1lekzkb"
                        >
                          <div
                            class="css-zvlevn"
                          >
                            Holder address
                          </div>
                        </div>
                      </th>
                    </tr>
                  </tbody>
                </table>
              </div>
              <div
                class="css-yzvkd8"
              >
                <p
                  class="css-0"
                >
                  No data
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
